File: game-server/src/main/java/work/fking/pangya/game/packet/outbound/ServerChannelsPacket.java
package work.fking.pangya.game.packet.outbound;

import io.netty.buffer.ByteBuf;
import work.fking.pangya.game.ServerChannel;
import work.fking.pangya.game.ServerChannel.Restriction;
import work.fking.pangya.networking.protocol.OutboundPacket;

import java.util.List;

import static work.fking.pangya.networking.protocol.ProtocolUtils.writeFixedSizeString;

public class ServerChannelsPacket implements OutboundPacket {

    private static final int ID = 0x4d;

    private final List<ServerChannel> serverChannels;

    public ServerChannelsPacket(List<ServerChannel> serverChannels) {
        this.serverChannels = serverChannels;
    }

    public static ServerChannelsPacket create(List<ServerChannel> serverChannels) {
        return new ServerChannelsPacket(serverChannels);
    }

    @Override
    public void encode(ByteBuf target) {
        target.writeShortLE(ID);
        target.writeByte(serverChannels.size());

        for (var channel : serverChannels) {
            writeFixedSizeString(target, channel.name(), 64);
            target.writeShortLE(channel.capacity());
            target.writeShortLE(channel.playerCount());
            target.writeShortLE(channel.id());
            target.writeShortLE(pack(channel.restrictions()));
            target.writeZero(5);
        }
    }

    private int pack(List<Restriction> restrictions) {
        var bitFlags = 0;
        for (var restriction : restrictions) {
            bitFlags |= restriction.flag();
        }
        return bitFlags;
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/packet/outbound/ServerChannelsPacket.kt
package work.fking.pangya.game.packet.outbound

import io.netty.buffer.ByteBuf
import work.fking.pangya.game.ServerChannel
import work.fking.pangya.game.ServerChannel.Restriction
import work.fking.pangya.networking.protocol.OutboundPacket
import work.fking.pangya.networking.protocol.writeFixedSizeString

class ServerChannelsPacket(
    private val serverChannels: List<ServerChannel>
) : OutboundPacket {

    override fun encode(buffer: ByteBuf) {
        buffer.writeShortLE(0x4d)
        buffer.writeByte(serverChannels.size)

        for (channel in serverChannels) {
            buffer.writeFixedSizeString(channel.name, 64)
            buffer.writeShortLE(channel.capacity)
            buffer.writeShortLE(channel.playerCount())
            buffer.writeShortLE(channel.id)
            buffer.writeShortLE(pack(channel.restrictions))
            buffer.writeZero(5)
        }
    }

    private fun pack(restrictions: List<Restriction>): Int {
        var bitFlags = 0
        for (restriction in restrictions) {
            bitFlags = bitFlags or restriction.flag()
        }
        return bitFlags
    }
}
